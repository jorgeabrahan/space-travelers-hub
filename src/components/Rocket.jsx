/* eslint-disable import/no-extraneous-dependencies */
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { cancelRocket, reserveRocket } from '../redux/rockets/rocketsSlice';

const Rocket = ({
  id, name, image, description, reserved,
}) => {
  const dispatch = useDispatch();
  const handleReserve = () => {
    dispatch(reserveRocket(id));
  };
  const handleCancelation = () => {
    dispatch(cancelRocket(id));
  };

  return (
    <div className="rocket">
      <img className="rocket__image" src={image} alt={name} loading="lazy" />
      <div className="rocket__content">
        <h2 className="rocket__title">{ name }</h2>
        <p className="rocket__description">
          {reserved && <span className="rocket__badge">Reserved</span>}
          {description}
        </p>
        {!reserved && (
        <button
          className="rocket__button reserve"
          type="button"
          onClick={handleReserve}
        >
          Reserve Rocket
        </button>
        )}
        {reserved && (
        <button
          className="rocket__button reserved"
          type="button"
          onClick={handleCancelation}
        >
          Cancel Reservation
        </button>
        )}
      </div>
    </div>
  );
};

Rocket.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};

export default Rocket;
