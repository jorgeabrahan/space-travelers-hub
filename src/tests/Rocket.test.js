/* eslint-disable import/no-extraneous-dependencies */
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Rocket from '../components/Rocket';
import store from '../redux/store';

describe('Tests for <Rocket /> component', () => {
  const title = 'Falcon 1';
  const titleNoSpace = title.replace(/\s/g, '');
  const description = 'Falcon description';
  const createRocket = (reserved = false) => (
    <Rocket
      id={titleNoSpace}
      name={title}
      image={`${titleNoSpace}.jpg`}
      description={description}
      reserved={reserved}
    />
  );

  const wrapper = (component) => (
    <Provider store={store}>
      {component}
    </Provider>
  );

  it('Should render the content correctly', () => {
    render(wrapper(createRocket()));
    // Title is rendering correctly
    expect(screen.getByText(title)).not.toBeNull();
    // Description is rendering correctly
    expect(screen.getByText('Falcon description')).not.toBeNull();
    // Rocket image is rendering correctly
    expect(screen.queryByRole('img', { src: `${titleNoSpace}.jpg` })).not.toBeNull();
  });

  it('Should display the reserve button when reserved is false', () => {
    render(wrapper(createRocket()));
    // Reserve button is rendering correctly
    expect(screen.queryByRole('button', { children: 'Reserve Rocket' })).not.toBeNull();
    // Cancel reservation button should not be displayed
    expect(() => screen.getByText('Cancel Reservation')).toThrow();
  });

  it('Should display the cancel reservation button when reserved is true', () => {
    render(wrapper(createRocket(true)));
    // Cancel reservation button is rendering correctly
    expect(screen.queryByRole('button', { children: 'Cancel Reservation' })).not.toBeNull();
    // Reserve button should not be displayed
    expect(() => screen.getByText('Reserve Rocket')).toThrow();
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(createRocket())).toJSON()).toMatchSnapshot();
  });
});
