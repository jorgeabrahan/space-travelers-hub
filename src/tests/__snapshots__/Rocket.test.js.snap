// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Rocket /> component Should match the snapshot 1`] = `
<div
  className="rocket"
>
  <img
    alt="Falcon 1"
    className="rocket__image"
    loading="lazy"
    src="Falcon1.jpg"
  />
  <div
    className="rocket__content"
  >
    <h2
      className="rocket__title"
    >
      Falcon 1
    </h2>
    <p
      className="rocket__description"
    >
      Falcon description
    </p>
    <button
      className="rocket__button reserve"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
